### PR TITLE
feat!: update terraform to version 1.6.x, relax max version constraint

### PR DESCRIPTION
**Pull request by johnstcn**
Relates to https://github.com/coder/coder/issues/12009

- Updates default Terraform version to 1.6.6
- Relaxes maximum version constraint on installed Terraform version

This should allow both us and others to experiment more easily with newer versions of Terraform.

Marking this as a breaking change so that it's called out prominently in the release notes.

### CODE REVIEW

File: scripts/Dockerfile.base [hunk 1]
```diff
@@ -26,9 +26,7 @@ RUN apk add --no-cache \
 # Terraform was disabled in the edge repo due to a build issue.
 # https://gitlab.alpinelinux.org/alpine/aports/-/commit/f3e263d94cfac02d594bef83790c280e045eba35
 # Using wget for now. Note that busybox unzip doesn't support streaming.
-#
-# WARNING: Do not update to 1.6.x, as it is the first release licensed under BSL instead of MPL.
```
Comment (deansheather) [inline scripts/Dockerfile.base:30]: We're OK with this from a legal standpoint?

**Review by deansheather — APPROVED** [review]

File: scripts/Dockerfile.base (hunk 1 above)
Comment (johnstcn) [reply scripts/Dockerfile.base:30]: > We're OK with this from a legal standpoint?

I'm assuming so based on @bpmct 's wording [here](https://github.com/coder/coder/issues/12009#issue-2117436345) but good to super confirm.

**Review by mtojek — APPROVED** [review]
I'm fine with this change, although it would be great to get 👍  from leadership on the legal side too.

File: provisioner/terraform/serve.go [hunk 2]
```diff
@@ -55,23 +55,38 @@ func absoluteBinaryPath(ctx context.Context) (string, error) {
 		return "", xerrors.Errorf("Terraform binary absolute path not found: %w", err)
 	}
 
-	// Checking the installed version of Terraform.
-	version, err := versionFromBinaryPath(ctx, absoluteBinary)
+	// Checking the installed installedVersion of Terraform.
```
Comment (johnstcn) [inline provisioner/terraform/serve.go:58]: over-zealous rename

```suggestion
	// Checking the installed version of Terraform.
```

**Comment by matifali** [comment]
Why not go to 1.7.x?

**Comment by johnstcn** [comment]
> Why not go to 1.7.x?

Our current default is 1.4.6, and going to 1.6.6 is already a big jump.
I'd honestly be happier jumping to 1.5.7 but that's already out of support.

File: scripts/Dockerfile.base (hunk 1 above)
Comment (coadler) [reply scripts/Dockerfile.base:30]: Yeah we're good

**Comment by coadler** [comment]
Yeah, I think jumping to 1.6.x first is the play here

**Review by coadler — APPROVED** [review]

**Comment by johnstcn** [comment]
I'm going to hold off on merging this until @bpmct gives the green light. I'd feel most comfortable doing this right after the next release so that we have a good deal of time to iron out any issues we run into on dogfood.

**Comment by matifali** [comment]
I am personally using 1.7.x on one of my deployments (coder Deb package and  tarraform installed manually) and so far no issues. But I agree we should do it after the release.

**Comment by matifali** [comment]
@johnstcn I have terraform 1.7.2 and Coder seems to work even though it's outside of the max version. Shouldn't the builds fail in this case?

**Comment by johnstcn** [comment]
> @johnstcn I have terraform 1.7.2 and Coder seems to work even though it's outside of the max version. Shouldn't the builds fail in this case?

It should download the 'known good' Terraform version in this case.

**Review by bpmct — APPROVED** [review]
LGTM! Approved

**Pull request merged**